src/main/java/org/elasticsearch/index/query/IdsQueryBuilder.java

```diff
@@ -159,11 +149,13 @@ public Query toQuery(QueryParseContext parseContext) throws IOException, QueryPa
             return Queries.newMatchNoDocsQuery();
         }
 
-        Collection<String> typesForQuery = this.types;
-        if (typesForQuery == null || typesForQuery.isEmpty()) {
+        Collection<String> typesForQuery;
+        if (types == null || types.length == 0) {
             typesForQuery = parseContext.queryTypes();
-        } else if (typesForQuery.size() == 1 && Iterables.getFirst(typesForQuery, null).equals("_all")) {
+        } else if (types.length == 1 && MetaData.ALL.equals(types[0])) {
             typesForQuery = parseContext.mapperService().types();
+        } else {
+            typesForQuery = Sets.newHashSet(types);
         }
 
         TermsQuery query = new TermsQuery(UidFieldMapper.NAME, Uid.createTypeUids(typesForQuery, ids));
```

Comment:
Theres a problem here when changing ids fro List<String> to Set<String> that I ran into when running the tests. The way Uid.createTypeUids handles this internally is that it will cast the Set to Object and then treat that as a one element list when creating the query. Not obvious from this line, but you can run SearchQueryTests#testBasicQueryById(). I don't know whats better, changing the code in Uid.createTypeUids or converting back to List here. I like the internal usage of Set for the ids, although I wonder if at some point we might need a fixed ordering of the elements.